tokio/src/task/scope.rs

```diff
@@ -0,0 +1,439 @@
+    drop_behavior: Option<ScopeDropBehavior>,
+}
+
+impl ScopeConfigBuilder {
+    /// Creates a new scope which is treated as a child scope of the given
+    /// [`ScopeHandle`]. The new scope will inherit all properties of the parent
+    /// scope. In addition tasks inside the new scope will get cancelled when
+    /// the parent scope gets cancelled.
+    pub fn with_parent(parent: ScopeHandle) -> Self {
+        Self {
+            parent: Some(parent),
+            drop_behavior: None,
+        }
+    }
+
+    /// Creates a new scope which is detached from any parent scope.
+    /// Tasks spawned on this `scope` will not get cancelled if any parent scope
+    /// gets cancelled. Instead those tasks would only get cancelled if the
+    /// scope itself gets cancelled.
+    pub fn detached() -> Self {
+        Self {
+            parent: None,
+            drop_behavior: Some(ScopeDropBehavior::Panic),
+        }
+    }
+
+    /// If the scope is dropped instead of being awaited, the thread which
+    /// performing the drop of the scope will block until all child tasks in the
+    /// scope will have run to completion.
+    pub fn block_to_completion(&mut self) {
+        self.drop_behavior = Some(ScopeDropBehavior::BlockToCompletion);
+    }
+
+    /// If the scope is dropped instead of being awaited, child tasks will
+    /// continue to run. This breaks "structured concurrency", since child tasks
+    /// are now able to outlive the parent task.
+    pub fn continue_tasks_on_drop(&mut self) {
+        self.drop_behavior = Some(ScopeDropBehavior::ContinueTasks);
+    }
+
+    /// If a scope `Future` gets dropped instead of awaited, the current process
+    /// will be aborted. This settings tries to provide higher guarantees about
+    /// child tasks not outliving their parent tasks.
+    pub fn abort_on_drop(&mut self) {
+        self.drop_behavior = Some(ScopeDropBehavior::Abort);
+    }
+    
+    /// If a scope `Future` gets dropped instead of awaited, the current
+    /// thread will `panic!` in order to indicate that the scope did not
+    /// correctly wait for child tasks to complete, and that there exist detached
+    /// child tasks as a result of this action.
+    pub fn panic_on_drop(&mut self) {
+        self.drop_behavior = Some(ScopeDropBehavior::Panic);
+    }
+
+    /// Builds the configuration for the scope
+    pub fn build(self) -> Result<ScopeConfig, ScopeConfigBuilderError> {
+        // Get defaults
+
+        // Generate a `CancellationToken`. If a parent scope and an associated
+        // cancellation token exists, we create a child token from it.
+        let cancellation_token = if let Some(parent) = &self.parent {
+            parent.scope.config.cancellation_token.child_token()
+        } else {
+            CancellationToken::new()
+        };
+
+        let mut drop_behavior = match &self.parent {
+            Some(parent) => parent.scope.config.drop_behavior,
+            None => ScopeDropBehavior::Panic,
```

Comment:
Should it really panic by default if you drop a scope without polling it to the end? I can imagine plenty of Rust patterns where you might start a task/scope at the beginning, and then try some other fallible method, and only afterwards `scope.await`. If you were using `?`, it'd be pretty easy to drop the scope, and probably most would expect that the scope would just be canceled.

Reply:
As written in a few other places: It's for obvious reasons not a great strategy. But there exists no perfect strategy for this problem, and therefore this seems to be the best compromise.

- letting the background task continue to run (as proposed in #2596 ) will break structured concurrency and will not allow users to reason anymore about when subtasks (and their usage for certain resources) end
- doing a blocking wait will not work in single-threaded environments, and might lead to high latency variances in multithreaded environments.